### PR TITLE
Fix memory struct allocation in Sol->Yul.

### PR DESCRIPTION
**Pull request by ekpyron**

### CODE REVIEW

File: test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    struct S {
+        uint256 a;
+        uint256 b;
+    }
+
+    function f() public pure returns (uint256 a, uint256 b){
+        assembly {
+            // Make free memory dirty to check that the struct allocation cleans it up again.
+            let freeMem := mload(0x40)
+            mstore(freeMem, 42)
```
Comment (bshastry) [inline test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol:11]: Note to self: Store `42` at memory address `0x80`

File: test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+            mstore(add(freeMem, 32), 42)
```
Comment (bshastry) [inline test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol:12]: Here we are storing `42` at memory address `0x80 + 0x20 = 0xa0`

File: test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        }
+        S memory s;
```
Comment (bshastry) [inline test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol:14]: Why is the struct allocation expected to clean up previous two mstores (as per the comment above)?

File: test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        return (s.a, s.b);
```
Comment (bshastry) [inline test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol:15]: Note to self: Since struct is uninitialized `s.a` and `s.b` default to `0`. Actually here we are testing that they are not `42` (previous value at locations where `s.a` and `s.b` reside)!

File: test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol (hunk 2 above)
Comment (bshastry) [reply test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol:12]: Note to self: We don't update free memory pointer to hold `0xa0 + 0x20 = 0xc0` so it still points to `0x80`

Comment (bshastry) [reply test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol:12]: So technically, we expect `S` to not hold dirty bits stored at addresses `0x80` and `0xa0` respectively `42` in both cases.

File: test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol (hunk 3 above)
Comment (bshastry) [reply test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol:14]: Ah I think I answered my question above :-)

tl;dr: Uninitialized allocation should clear previously stored memory values

**Review by bshastry — APPROVED** [review]
Looks good to me.

File: test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol (hunk 4 above)
Comment (ekpyron) [reply test/libsolidity/semanticTests/viaYul/memory_struct_allow.sol:15]: Exactly :-).
